[package] add mock/extern hooks

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57341 [package] Add an intern keyword
* **#58000 [package] add mock/extern hooks**

Directly overriding save_extern and save_mock may mess with our
invariants in weird ways. This is less pronounced now, but once we
switch to graph-based dependency management things will get broken
subtly if people fail to call `super()`.

Better to add hook support to reflect that really you can only do a side
effect. Also has the bonus that people are likely familiar with it from
`nn.Module` hooks.

Differential Revision: [D28339191](https://our.internmc.facebook.com/intern/diff/D28339191)

suo committed:
[package] add mock/extern hooks

Directly overriding save_extern and save_mock may mess with our
invariants in weird ways. This is less pronounced now, but once we
switch to graph-based dependency management things will get broken
subtly if people fail to call `super()`.

Better to add hook support to reflect that really you can only do a side
effect. Also has the bonus that people are likely familiar with it from
`nn.Module` hooks.

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28339191).

suo commented on test/package/test_misc.py:
```diff
@@ -139,31 +137,6 @@ def test_is_from_package(self):
         self.assertTrue(is_from_package(loaded_obj))
 
-
-    @skipIf(version_info < (3, 7), "mock uses __getattr__ a 3.7 feature")
```
I am deleting this test because I want to remove the `require_module` extension point, which has a similar "may bust our invariants" issue. I think `require_module` was more useful before we had robust pattern matching.

suo committed:
Update on "[package] add mock/extern hooks"


Directly overriding save_extern and save_mock may mess with our
invariants in weird ways. This is less pronounced now, but once we
switch to graph-based dependency management things will get broken
subtly if people fail to call `super()`.

Better to add hook support to reflect that really you can only do a side
effect. Also has the bonus that people are likely familiar with it from
`nn.Module` hooks.

Differential Revision: [D28339191](https://our.internmc.facebook.com/intern/diff/D28339191)

[ghstack-poisoned]

suo committed:
Update on "[package] add mock/extern hooks"


Directly overriding save_extern and save_mock may mess with our
invariants in weird ways. This is less pronounced now, but once we
switch to graph-based dependency management things will get broken
subtly if people fail to call `super()`.

Better to add hook support to reflect that really you can only do a side
effect. Also has the bonus that people are likely familiar with it from
`nn.Module` hooks.

Differential Revision: [D28339191](https://our.internmc.facebook.com/intern/diff/D28339191)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28339191).

SplitInfinity commented on test/package/test_dependency_hooks.py:
```diff
@@ -0,0 +1,101 @@
+            exporter.register_extern_hook(my_extern_hook)
+            exporter.save_source_string("foo", "import module_a")
+
+        self.assertEqual(my_externs, set(["module_a"]))
```
```suggestion
        self.assertSetEqual(my_externs, set(["module_a"]))
```

SplitInfinity commented on test/package/test_dependency_hooks.py:
```diff
@@ -0,0 +1,101 @@
+            exporter.register_extern_hook(my_extern_hook2)
+            exporter.save_source_string("foo", "import module_a")
+
+        self.assertEqual(my_externs, set())
```
```suggestion
        self.assertSetEqual(my_externs, set())
```

SplitInfinity commented on test/package/test_dependency_hooks.py:
```diff
@@ -0,0 +1,101 @@
+
+        self.assertEqual(my_externs, set())
+
+    def test_remove_hooks(self):
```
Consider adding tests for multiple mock hooks.

SplitInfinity commented on test/package/test_dependency_hooks.py:
```diff
@@ -0,0 +1,101 @@
+
+        self.assertEqual(my_externs, set(["module_a"]))
+
+    def test_multiple_hooks(self):
```
Consider adding a test for multiple mock hooks.

SplitInfinity commented on test/package/test_dependency_hooks.py:
```diff
@@ -0,0 +1,101 @@
+            exporter.register_mock_hook(my_mock_hook)
+            exporter.save_source_string("foo", "import module_a; import package_a")
+
+        self.assertEqual(my_externs, set(["module_a"]))
+        self.assertEqual(my_mocks, set(["package_a"]))
```
```suggestion
        self.assertSetEqual(my_externs, set(["module_a"]))
        self.assertSetEqual(my_mocks, set(["package_a"]))
```

SplitInfinity reviewed (commented):
So for module hooks, I *think* there is a guarantee that they are invoked in the same order that they are registered. I think the same might be true for mock/extern hooks based on the fact that they are stored in `OrderedDicts` (although that data structure was chosen for compatibility with `RemovableHandle`), but the tests are written as if there is no such guarantee, and there is no mention of the invocation order in the documentation.

suo committed:
Update on "[package] add mock/extern hooks"


Directly overriding save_extern and save_mock may mess with our
invariants in weird ways. This is less pronounced now, but once we
switch to graph-based dependency management things will get broken
subtly if people fail to call `super()`.

Better to add hook support to reflect that really you can only do a side
effect. Also has the bonus that people are likely familiar with it from
`nn.Module` hooks.

Differential Revision: [D28339191](https://our.internmc.facebook.com/intern/diff/D28339191)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28339191).

suo committed:
Update on "[package] add mock/extern hooks"


Directly overriding save_extern and save_mock may mess with our
invariants in weird ways. This is less pronounced now, but once we
switch to graph-based dependency management things will get broken
subtly if people fail to call `super()`.

Better to add hook support to reflect that really you can only do a side
effect. Also has the bonus that people are likely familiar with it from
`nn.Module` hooks.

Differential Revision: [D28339191](https://our.internmc.facebook.com/intern/diff/D28339191)

[ghstack-poisoned]
